src/GeniusYield/TxBuilder/IO.hs

```diff
@@ -0,0 +1,104 @@
+{-|
+Module      : GeniusYield.TxBuilder.IO
+Copyright   : (c) 2023 GYELD GMBH
+License     : Apache 2.0
+Maintainer  : support@geniusyield.co
+Stability   : develop
+
+-}
+module GeniusYield.TxBuilder.IO (
+    GYTxMonadIO,
+    GYTxQueryMonadIO,
+    GYTxBuilderMonadIO,
+    runGYTxBuilderMonadIO,
+    runGYTxQueryMonadIO,
+    runGYTxMonadIO,
+    queryAsBuilderMonad,
+    liftQueryMonad,
+    liftBuilderMonad
+) where
+
+
+import           Control.Monad.Reader            (ReaderT(ReaderT), MonadReader, asks)
+import           Data.Maybe                      (maybeToList)
+
+import           GeniusYield.TxBuilder.Class
+import           GeniusYield.TxBuilder.Errors
+import           GeniusYield.TxBuilder.IO.Query
+import           GeniusYield.TxBuilder.IO.Builder
+import           GeniusYield.Types
+
+-------------------------------------------------------------------------------
+-- GY implementation
+-------------------------------------------------------------------------------
+
+-- | 'GYTxMonad' interpretation run under IO.
+type role GYTxMonadIO representational
+newtype GYTxMonadIO a = GYTxMonadIO (GYTxIOEnv -> GYTxBuilderMonadIO a)
+  deriving ( Functor
+           , Applicative
+           , Monad
+           , MonadReader GYTxIOEnv
+           , MonadRandom
+           , MonadError GYTxMonadException
+           , GYTxQueryMonad
+           , GYTxSpecialQueryMonad
+           , GYTxUserQueryMonad
+           )
+  via ReaderT GYTxIOEnv GYTxBuilderMonadIO
+
+data GYTxIOEnv = GYTxIOEnv
+    { envNid           :: !GYNetworkId
+    , envProviders     :: !GYProviders
+    , envPaymentSKey   :: !GYPaymentSigningKey
+    , envStakeSKey     :: !(Maybe GYStakeSigningKey)
```

Comment:
It now contains the user's payment and stake signing key.

Note that this monad is likely not necessary for app-like usecases. E.g the GY app code should probably use `GYTxBuilderMonadIO` instead. Also exported here.